### PR TITLE
[Frontend][TFlite] Add parser support for relu6, leaky_relu, relu_n1_to_1, log_softmax

### PR DESCRIPTION
**Pull request by inadob**
**This patch is made on top of #4789.** 
* Add helper functions in the parser to quantize and dequantize with Relay
* Add qnn tests for all 4 activation functions

### CODE REVIEW

File: python/tvm/relay/frontend/tflite.py [hunk 1]
```diff
@@ -463,6 +484,114 @@ def convert_relu(self, op):
...
+        if input_tensor.qnn_params:
+            in_expr = self.dequantize(in_expr, input_tensor)
+        out = _op.clip(in_expr, a_min=0, a_max=6)
+        if output_tensor.qnn_params:
+            out = self.quantize(out, output_tensor)
+
+        return out
+
+    def convert_leaky_relu(self, op):
+        """Convert TFLite LEAKY_RELU"""
```
Comment (wyc-ruiker) [inline python/tvm/relay/frontend/tflite.py]: Not important, I think it should be `"""Convert TFLite Leaky_ReLU"""` to align with `"""One iteration of Leaky_ReLU"""`.

File: python/tvm/relay/frontend/tflite.py [hunk 2]
```diff
@@ -463,6 +484,114 @@ def convert_relu(self, op):
...
+        if input_tensor.qnn_params:
+            in_expr = self.dequantize(in_expr, input_tensor)
+        out = _op.nn.leaky_relu(in_expr, alpha_tensor)
+        if output_tensor.qnn_params:
+            out = self.quantize(out, output_tensor)
+
+        return out
+
+    def convert_relu_n1_to_1(self, op):
+        """Convert TFLite RELU_N1_TO_1"""
```
Comment (wyc-ruiker) [inline python/tvm/relay/frontend/tflite.py]: Not important, I think it should be `"""Convert TFLite ReLU_n1_to_1"""` to align with `"""One iteration of ReLU_n1_to_1"""`.

**Comment by inadob** [comment]
> Relu and Clip implementation does not look right.
> 
> We can keep the computation in integer domain. The way to do that is to subtract the input zero point, and then call Relu, then requantize to the output scale (only if output scale/zero point are different from input scale/zero point).

Is that lowering to int8, not int32?

**Comment by tqchen** [comment]
@inadob please followup to address the review comments :)

**Comment by inadob** [comment]
> Relu and Clip implementation does not look right.
> 
> We can keep the computation in integer domain. The way to do that is to subtract the input zero point, and then call Relu, then requantize to the output scale (only if output scale/zero point are different from input scale/zero point).

That's fine for the standard ReLU op but I am not entirely sure whether we can follow this logic when we apply ReLU6 and ReLU1. Is there a way to recreate them in Relay without clip?.

What I did, in this case, was to shift the data by subtracting the zero point, do clip, shift back and finally requantize if needed. The problem here was that `Clip()` from `RELAY_PASS_PATTERN ` needs the clipping range to be in `double` and I couldn't just fix this by casting to `float64` (it seems `double` is not supported in TVM).

@anijain

**Comment by anijain2305** [comment]
> The problem here was that `Clip()` from `RELAY_PASS_PATTERN ` needs the clipping range to be in `double` and I couldn't just fix this by casting to `float64` (it seems `double` is not supported in TVM).

Can you show me code example here? I have typically used float numbers like 1.0 or 6.0 to work with `a_min` or `a_max`

**Comment by inadob** [comment]
> Can you show me code example here? I have typically used float numbers like 1.0 or 6.0 to work with `a_min` or `a_max`

```c++
Expr ClipQnnCanonicalize(const Attrs& attrs, const Array<Expr>& new_args,
                               const Array<tvm::relay::Type>& arg_types) {
  CHECK_EQ(new_args.size(), 7);
  auto& input_tensor = new_args[0];
  auto& input_scale = new_args[1]; // in fp
  auto& input_zero_point = new_args[2]; // in int32
  auto& clip_min = new_args[3];  // value is in fp
  auto& clip_max = new_args[4];  // value is in fp
  auto& output_scale = new_args[5];
  auto& output_zero_point = new_args[6];

  // Get the input dtype and shape.
  CHECK_EQ(arg_types.size(), 8);
  auto tensor_type = arg_types[0].as<TensorTypeNode>();
  CHECK(tensor_type != nullptr);
  auto input_dtype = tensor_type->dtype;
  auto input_shape = tensor_type->shape;


  // shift the input by subtracting the input zero_point
  auto shifted_input = Subtract(Cast(input_tensor, DataType::Int(32)), input_zero_point);

  // do the clipping in int32
  // auto clipped_tensor = Clip(shifted_input, clip_min, clip_max)
  auto clipped_tensor = Clip(shifted_input, Cast(clip_min, DataType::Float(64)), Cast(clip_max, DataType::Float(64)))
  // shift the input back by adding the zero_point
  clipped_tensor = Add(clipped_tensor, input_zero_point);

  // requantize the output if needed
  auto requantized_output = clipped_tensor;
  if (!IsEqualScalar(input_scale, output_scale) ||
      !IsEqualScalar(input_zero_point, output_zero_point)) {
     requantized_output = Requantize(clipped_tensor, input_shape, input_scale, input_zero_point, output_scale,
                                     output_zero_point, DataType::Int(32));
    }

  // Go back to lower precision.
  auto q_min = GetQmin(input_dtype);
  auto q_max = GetQmax(input_dtype);
  requantized_output = Clip(requantized_output, q_min, q_max); 
  return Cast(requantized_output, input_dtype);
}
```
And I am getting a complaint about **double dtype**...
```
/workspace/src/relay/qnn/op/clip.cc:61:117: error: cannot convert 'tvm::relay::Expr {aka tvm::RelayExpr}' to 'double' for argument '2' to 'tvm::relay::Expr tvm::relay::Clip(tvm::relay::Expr, double, double)'
   auto clipped_tensor = Clip(shifted_input, Cast(clip_min, DataType::Float(64)), Cast(clip_max, DataType::Float(64)))

```

The commented out line where I directly do clip() without casting to float64 didn't work too.

**Comment by tqchen** [comment]
@anijain2305 @inadob please followup :)

**Comment by tqchen** [comment]
ping

**Comment by inadob** [comment]
@anijain can you please help me with the quantized operators

**Comment by anijain2305** [comment]
Sorry for the delay. @inadob Can you take a look at this PR - https://github.com/apache/incubator-tvm/pull/5479

I have added support for quantized activations there. We dont need to create a new op in C++ from scratch. It is pretty easy to do the same thing in python. Let me know what you think

File: python/tvm/relay/frontend/tflite.py [hunk 3]
```diff
@@ -463,6 +484,114 @@ def convert_relu(self, op):
...
+
+        output_tensors = self.get_output_tensors(op)
+        assert len(output_tensors) == 1, "output tensors length should be 1"
+        output_tensor = output_tensors[0]
+
+        if input_tensor.qnn_params:
+            in_expr = self.dequantize(in_expr, input_tensor)
+        out = _op.clip(in_expr, a_min=0, a_max=6)
+        if output_tensor.qnn_params:
+            out = self.quantize(out, output_tensor)
```
Comment (anijain2305) [inline python/tvm/relay/frontend/tflite.py:508]: This is expensive. Lets try this https://github.com/apache/incubator-tvm/pull/5479/files

**Comment by anijain2305** [comment]
Ping @inadob Let us know if you are working on this. Or else I can take a chance at this.

**Comment by inadob** [comment]
> Ping @inadob Let us know if you are working on this. Or else I can take a chance at this.
@anijain2305 Yes, I am working on it - I will be ready in the next few days. Sorry for the delay

**Comment by anijain2305** [comment]
@inadob Your changes look better now. Can you please rebase? (`git submodule update --init --recursive`)

**Comment by anijain2305** [comment]
It seems you still have old `3rdparty/dmlc-core'. You can check that by clicking on "Files changed" tab.

**Comment by inadob** [comment]
> It seems you still have old `3rdparty/dmlc-core'. You can check that by clicking on "Files changed" tab.

It's fixed now.

**Comment by anijain2305** [comment]
@inadob Can you please fix the CI error?

**Comment by anijain2305** [comment]
Ping @inadob :)

**Comment by tqchen** [comment]
Glad that it is moving, @anijain2305 please manage this PR

**Comment by inadob** [comment]
> Ping @inadob :)

Fixed it. :)

**Comment by inadob** [comment]
@anijain2305 
CI has been fixed :)

**Comment by anijain2305** [comment]
Thanks @inadob This is merged!

**Pull request merged**